src/network_service/sending_targets_cache.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright 2019 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use crate::ConnectionInfo;
+use log::LogLevel;
+use std::collections::HashMap;
+use std::net::SocketAddr;
+
+const MAX_RESENDS: u8 = 3;
+
+enum TargetState {
+    Sending(u8),
+    Failed(u8),
+    Sent,
+}
+
+impl TargetState {
+    pub fn is_complete(&self) -> bool {
+        match *self {
+            TargetState::Failed(x) => x > MAX_RESENDS,
+            TargetState::Sent => true,
+            TargetState::Sending(_) => false,
+        }
+    }
+
+    pub fn is_sending(&self) -> bool {
+        match *self {
+            TargetState::Failed(_) | TargetState::Sent => false,
+            TargetState::Sending(_) => true,
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct SendingTargetsCache {
+    cache: HashMap<u64, Vec<(ConnectionInfo, TargetState)>>,
+}
+
+impl SendingTargetsCache {
+    pub fn insert_message(
+        &mut self,
+        msg_id: u64,
+        initial_targets: Vec<ConnectionInfo>,
+        dg_size: usize,
+    ) {
+        let targets = initial_targets
+            .into_iter()
+            .enumerate()
+            .map(|(idx, tgt_info)| {
+                (
+                    tgt_info,
+                    if idx < dg_size {
```

Comment:
could you pls explain why this works? thx

Reply:
This sets `dg_size` initial targets' states to `Sending`, and the rest to `Failed(0)`. This is because we initially only send to the first `dg_size` of potential targets. `idx` is just an index we get via `enumerate()`, so it grows from 0 upwards and exactly the first `dg_size` items will have the index less than `dg_size`.